### PR TITLE
Enhance post-processing module performance

### PR DESCRIPTION
**Pull request by voferreira**
# Description of the problem

- Lethe_pyvista_tools was too slow. My routine for 35-second cases with 100k particles took more than 57 minutes to post-process.

- Some members would have unnecessary repeated calls.

# Description of the solution

- Made the module installable, which helped improve performance.
- It is more like real OOP now, which allows reducing module imports and repeated code.
- Created some instances like ``self.sorted`` to avoid repetition.
- Applied multiprocessing on the parallelization of routines. Now running the code in parallel is as simple as it was before.
- Modified a few data structures here and there.

- Now it takes about 11 minutes on 6 cores.

# How Has This Been Tested?

- My entire post-processing routine works well.
- [x] Examples

# Documentation

- Created a section in tools for it;
- Updated the small-scale rotating drum post-processing example

# Future changes

- It needs tests to prevent it from getting too messy and keep track of changes.

# Comments

- This pr fixes ``__doc__`` using ``"""`` instead of ``'''``. 
- I also tried to use Cython to compile code as C, but it did not improve the performance.

### CODE REVIEW

**Comment by blaisb** [comment]
@voferreira can you request two reviewers other than me? You can ask on the Slack lethe channel :)

**Comment by voferreira** [comment]
> @voferreira can you request two reviewers other than me? You can ask on the Slack lethe channel :)
> 

Sure thing! I was asking people individually but I think the channel is more efficient haha.

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (blaisb) [inline doc/source/tools/postprocessing/postprocessing.rst:7]:
```suggestion
The module is powered by `PyVista <https://docs.pyvista.org/>`_, a powerful 3D plotting and mesh analysis tool. It is a "pythonic" interface to deal wih Visualization Toolkit (VTK) data. The module also uses the powerful `multiprocessing <https://docs.python.org/3/library/multiprocessing.html>`_ Python library to increase post-processing speed by running tasks in multiple processors.

```
use italic instead of " " for pythonic

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (blaisb) [inline doc/source/tools/postprocessing/postprocessing.rst:17]:
```suggestion
There are two ways to have access to ``lethe_pyvista_tools``: you can install the module and have it on your machine just like ``os``, ``sys``, or other PyPi modules; or, you can use it without installing. Here, we explain how to do both.

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (blaisb) [inline doc/source/tools/postprocessing/postprocessing.rst:22]:
```suggestion
Installing ``lethe_pyvista_tools`` is very simple. In the `contrib/postprocessing <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_ of your Lethe version you will find all source files, including ``setup.py`` and ``requirements.txt``. Those two files take care of configuring your installation and assuring that everything is set. To install it, navigate to `contrib/postprocessing <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_ on your terminal and run:

```

**Review by blaisb — CHANGES_REQUESTED** [review]
Few comments, good work!

File: contrib/postprocessing/lethe_pyvista_tools/__init__.py
Comment (voferreira) [reply contrib/postprocessing/lethe_pyvista_tools/__init__.py:9]: Not necessarily. I am doing this because the only class is this one and it has the same name as the folder. But If we had more object types I would maybe create a separate object file or multiple folders.

File: contrib/postprocessing/lethe_pyvista_tools/_modify_array.py
Comment (voferreira) [reply contrib/postprocessing/lethe_pyvista_tools/_modify_array.py:88]: A global method. It needs to be global to use multiprocessing. I couldn't make it work otherwise.

File: doc/source/examples/multiphysics/air-bubble-compression/air-bubble-compression.rst
Comment (voferreira) [reply doc/source/examples/multiphysics/air-bubble-compression/air-bubble-compression.rst:308]: Me neither. I just did not think of a better way to make sure the postprocessing file is working. I suppose users are not necessarily going to launch the simulations and the post-processing code from within lethe's folder, and the way most .py post-processing files are written kind of requires this (like ``path_to_module = '../../contrib/postprocessing/'``).

I can fix it by referring to $LETHE_PATH like it is done on the bouncing particles example (``path_to_module = '$LETHE_PATH/contrib/postprocessing/'``). But I am not sure if the user looking at an isolated example will catch everything needed to run it, especially given the new multiprocessing default behavior of running on all machine processors. Maybe I could change the default to 1 processor instead, but the dependencies would still be a requirement out of sight without the warning.

Let me know your final decision.

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst
Comment (lpsaavedra) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:15]:
```suggestion
  Details about installing the module or using it without installing it are available on this `documentation <../../../tools/postprocessing/postprocessing.py>`_.
```

File: doc/source/examples/unresolved-cfd-dem/liquid-solid-fluidized-bed/liquid-solid-fluidized-bed.rst
Comment (lpsaavedra) [inline doc/source/examples/unresolved-cfd-dem/liquid-solid-fluidized-bed/liquid-solid-fluidized-bed.rst:432]:
```suggestion
    This example includes a post-processing file written in Python that uses the `lethe_pyvista_tools <../../../tools/postprocessing/postprocessing.html>`_. module.
```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:7]:
```suggestion
The module is powered by `PyVista <https://docs.pyvista.org/>`_, a powerful 3D plotting and mesh analysis tool. It is a *pythonic* interface to deal with Visualization Toolkit (VTK) data. The module also uses the powerful `multiprocessing <https://docs.python.org/3/library/multiprocessing.html>`_ Python library to increase post-processing speed by running tasks in multiple processors.

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:12]:
```suggestion
  For `lethe_pyvista_tools <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_ to work, along with `Python 3 <https://www.python.org/downloads/>`_, the following libraries are needed: `os <https://docs.python.org/3/library/os.html>`_, `NumPy <https://numpy.org/>`_, `PyVista <https://docs.pyvista.org/>`_, `tqdm <https://tqdm.github.io/>`_, `matplotlib <https://matplotlib.org/stable/index.html>`_, `SciPy <https://scipy.org/>`_, and `scikit-learn <https://scikit-learn.org/stable/index.html>`_. If any of the modules are missing, use `pip <https://pypi.org/project/pip/>`_ to install it by running ``pip3 install $NAME_OF_THE_MODULE`` on the terminal. Alternatively, use the ``requirements.txt`` file and install them all at once running ``pip install -r requirements.txt`` located at ``contrib/postprocessing/``.

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:24]:
```suggestion
Installing ``lethe_pyvista_tools`` is very simple. In the `contrib/postprocessing <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_ of your Lethe version you will find all source files, including ``setup.py`` and ``requirements.txt``. Those two files take care of configuring your installation and assuring that everything is set correctly. To install it, navigate to `contrib/postprocessing <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_ on your terminal and run:

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:43]:
```suggestion
    Uninstalling ``lethe_pyvista_tools`` is exactly like doing so with any other library: ``pip3 uninstall lethe_pyvista_tools`` or ``pip3 uninstall lethe_pyvista_tools``.

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:48]: It may be trivial, but I think you could specify that you have to be in a python session, I'm not sure if python session is the correct expression, but I cannot think of anything better at the moment:
```suggestion
In this case, we use the `sys <https://docs.python.org/3/library/sys.html>`_ module to import it directly from Lethe's directory into a python session:

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:58]:
```suggestion
where `sys.path <https://docs.python.org/3/library/sys.html#sys.path:~:text=in%20version%203.10.-,sys.path%C2%B6,-A%20list%20of>`_ is a list of strings that specifies the search path for modules. Another option is to simply copy the `lethe_pyvista_tools.py <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing>`_ file to the same folder as the Python post-processing script, and import it as follows:

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:64]:
```suggestion
One third and very convenient way to always import the module without copying it or even adding the ``sys.path.append(path_to_module)`` is permanently adding the path to the module (i.e., ``/contrib/postprocessing/``) to your `PYTHONPATH <https://docs.python.org/3/library/sys_path_init.html#:~:text=The%20PYTHONPATH%20environment%20variable%20is,all%20installed%20Python%20versions%2Fenvironments.>`_.

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:66]: This should be moved right after the last code block, no? Before the "One third and very convenient...)

```suggestion
The ``*`` means that we want to import all members of lethe_pyvista_tools.

```

File: doc/source/tools/postprocessing/postprocessing.rst
Comment (lpsaavedra) [inline doc/source/tools/postprocessing/postprocessing.rst:74]:
```suggestion
To get quick-started, follow the hand-on `Small Scale Rotating Drum Post-processing example <../../examples/dem/small-scale-rotating-drum-post-processing>`_. It has a detailed explanation on how to use the module. You can also start with a "raw" template file, check `this example file <https://github.com/lethe-cfd/lethe/tree/master/contrib/postprocessing/example.py>`_.

```

**Review by lpsaavedra — CHANGES_REQUESTED** [review]
Good work! I think the Lethe PyVista Tools module is very readable and easy to follow. I have a couple of comments; most of them related to documentation. Apart from those, I also noticed that the `contrib/postprocessing/README.md` file needs to be updated before merging this PR.

File: contrib/postprocessing/lethe_pyvista_tools/__init__.py
Comment (voferreira) [reply contrib/postprocessing/lethe_pyvista_tools/__init__.py:278]: It is specific to this function. It is a way of saying that it is accessible to the other members.

File: contrib/postprocessing/lethe_pyvista_tools/__init__.py
Comment (voferreira) [reply contrib/postprocessing/lethe_pyvista_tools/__init__.py:277]: It is! Just fixed it!

**Review by blaisb — APPROVED** [review]
Seems good to me

File: doc/source/theory/fluid_dynamics/stabilization.rst
Comment (voferreira) [reply doc/source/theory/fluid_dynamics/stabilization.rst:13]: I fixed a small tipo on it. It is written "Petro-Galerkin"

**Review by lpsaavedra — APPROVED** [review]

File: contrib/postprocessing/lethe_pyvista_tools/_modify_array.py
Comment (PierreLaurentinCS) [inline contrib/postprocessing/lethe_pyvista_tools/_modify_array.py:30]:
```suggestion
    given condition), a numpy array or python list (with the same len of all

```

Comment (PierreLaurentinCS) [inline contrib/postprocessing/lethe_pyvista_tools/_modify_array.py:30]:
```suggestion
    given condition), an numpy array or python list (with the same length as all

```

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst
Comment (PierreLaurentinCS) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:65]:
```suggestion
Here, the constructor ``lethe_pyvista_tools`` receives three arguments: ``case_path``, ``prm_file_name``, and the name of the ``.pvd`` file generated by the simulation. In the above code line, ``"."`` means that the ``case_path`` is the path where we currently are, but it can be any path where the case is. The ``prm_file_name`` argument takes ``"small-rotating-drum-dem.prm"``. The constructor can take 3 other arguments: ``first``, ``last``, and ``step``, standing for the first and last time-steps to be read and the interval between the time-steps, respectively. By default, ``first = 0``, ``step = 1``, and ``last`` is the last time-step of the data.
```

File: doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst
Comment (PierreLaurentinCS) [inline doc/source/examples/dem/small-scale-rotating-drum-post-processing/small-scale-rotating-drum-post-processing.rst:69]:
```suggestion
    The special ``n_procs`` parameter controls the number of cores used in the post-processing routine. By default, it is set to ``None``. If no other value is provided, the ``n_procs`` will be the number of CPUs on your machines.
```

File: doc/source/examples/multiphysics/air-bubble-compression/air-bubble-compression.rst
Comment (voferreira) [reply doc/source/examples/multiphysics/air-bubble-compression/air-bubble-compression.rst]: @blaisb do you have any preference, or I leave it as is?

File: contrib/postprocessing/lethe_pyvista_tools/_get_cylindrical_coords.py
Comment (voferreira) [reply contrib/postprocessing/lethe_pyvista_tools/_get_cylindrical_coords.py:57]: arctan receives the angle as argument, while arctan2 receives y and x (respectively) and automatically applies trigonometry.

Comment (voferreira) [reply contrib/postprocessing/lethe_pyvista_tools/_get_cylindrical_coords.py:57]: https://numpy.org/doc/stable/reference/generated/numpy.arctan.html#numpy.arctan
https://numpy.org/doc/stable/reference/generated/numpy.arctan2.html

**Comment by blaisb** [comment]
will merge after CI :)

**Comment by voferreira** [comment]
> will merge after CI :)

The only comment missing is this one: https://github.com/lethe-cfd/lethe/pull/865#discussion_r1315699043

File: doc/source/examples/multiphysics/air-bubble-compression/air-bubble-compression.rst
Comment (blaisb) [reply doc/source/examples/multiphysics/air-bubble-compression/air-bubble-compression.rst]: leave as is

**Pull request merged**
